### PR TITLE
Use chunk hash for bundle generation for proper caching

### PR DESCRIPTION
**Pull request by uniibu**
This PR changes how bundle.js is generated. Closes https://github.com/theia-ide/theia/issues/3393

Currently, theia sets the filename statically to `bundle.js` and this prevents proper caching and `cache-busting` to be implemented.

By using `chunkhash` as the filename allows proper caching since the `chunkhash` changes every time a new code is added upon building. And with the use of [html-webpack-plugin](https://webpack.js.org/plugins/html-webpack-plugin/), it automatically injects the generated file to the `index.html`

This also allows to set a `max-age` to `1d` which tells cdn services to cache the resources/artifacts for 1 day unless the `chunkhash` has changed.

Signed-off-by: Uni Sayo <unibtc@gmail.com>

### CODE REVIEW

File: dev-packages/application-manager/src/generator/webpack-generator.ts [hunk 1]
```diff
@@ -54,7 +55,7 @@ const monacoHtmlLanguagePath = '${this.resolve('monaco-html', 'release/min')}';`
 module.exports = {
     entry: path.resolve(__dirname, 'src-gen/frontend/index.js'),
     output: {
-        filename: 'bundle.js',
+        filename: '[chunkhash].js',
```
Comment (davidfestal) [inline dev-packages/application-manager/src/generator/webpack-generator.ts]: Wouldn't it be better and a bit more clear to keep the `bundle` name before the hash suffix, like this:
```
bundle-[chunkhash].js
```
?

Comment (uniibu) [reply dev-packages/application-manager/src/generator/webpack-generator.ts]: You are right, I just thought that its too long. However, i think its preferrable to slice the `chunkhash` length to like 8. So it becomes 
```
bundle-[chunkhash:8].js
```
This way it is much clearer specially when debugging.

**Comment by svenefftinge** [comment]
@uniibu could you introduce a command line argument, that would optionally enable this for you?

**Comment by uniibu** [comment]
@svenefftinge added optional `--env.hashed` (defaults to false) to be used during `theia build`. Further info is on the commit message. Cheers!

**Comment by vince-fugnitto** [comment]
@uniibu do you still plan on updating the pull request and resolving the conflicts?
(I noticed the pull request is quite old, and may require additional review)

**Comment by akosyakov** [comment]
@vince-fugnitto I don't think there is an agreement to have it by default on. If it not on by default on then it can be done via the custom webpack configuration without additional options.

**Comment by vince-fugnitto** [comment]
> @vince-fugnitto I don't think there is an agreement to have it by default on. If it not on by default on then it can be done via [the custom webpack configuration](https://github.com/eclipse-theia/theia/issues/766#issuecomment-341720642) without additional options.

Does that mean that the pull request can be closed?
I haven't looked at the details of the pull request, I just noticed it's been in review a while and that there are currently conflicts.

**Comment by akosyakov** [comment]
yes, i think we can close it, if someone wants to enable something like that the PR is still in GitHub history

**Pull request closed without merging**